packages/compass-crud/src/components/virtualized-document-list-view.tsx

```diff
@@ -0,0 +1,127 @@
+  const occupiedStaticSizePerItem = spacing[400] + spacing[400];
+  // total size occupied by the "Show more fields" button whenever visible
+  const showMoreButtonSize =
+    DEFAULT_VISIBLE_FIELDS < doc.elements.size
+      ? 38 //spacing[400] + spacing[400]
+      : 0;
+  // total size occupied by initially rendered fields
+  const visibleFieldsSize =
+    Math.min(DEFAULT_VISIBLE_FIELDS, doc.elements.size) *
+    (spacing[400] + spacing[25]);
+
+  return (
+    keylineCardBorder +
+    occupiedStaticSizePerItem +
+    showMoreButtonSize +
+    visibleFieldsSize
+  );
+};
+
+type VirtualizedDocumentListViewProps = {
+  docs: (HadronDocument | BSONObject)[];
+  isEditable: boolean;
+  className?: string;
+  initialScrollTop?: number;
+  scrollTriggerRef?: React.Ref<HTMLDivElement>;
+  scrollableContainerRef?: React.Ref<HTMLDivElement>;
+} & Pick<
+  DocumentProps,
+  | 'isTimeSeries'
+  | 'copyToClipboard'
+  | 'removeDocument'
+  | 'replaceDocument'
+  | 'updateDocument'
+  | 'openInsertDocumentDialog'
+>;
+
+const VirtualizedDocumentListView: React.FC<
+  VirtualizedDocumentListViewProps
+> = ({
+  docs: _docs,
+  isEditable,
+  className,
+  isTimeSeries,
+  initialScrollTop,
+  scrollTriggerRef,
+  scrollableContainerRef,
+  copyToClipboard,
+  removeDocument,
+  replaceDocument,
+  updateDocument,
+  openInsertDocumentDialog,
+}) => {
+  const docs = useMemo(() => {
+    return _docs.map((_doc) => {
+      // COMPASS-5872 If doc is a plain js object rather than an instance of hadron-document Document
+      // it may have an isRoot prop, which would cause the isRoot() to throw an error.
+      if (typeof _doc?.isRoot === 'function' && _doc?.isRoot()) {
+        return _doc as HadronDocument;
+      }
+      return new HadronDocument(_doc as any);
+    });
+  }, [_docs]);
+
+  const renderItem: VirtualListItemRenderer<HadronDocument> = useCallback(
+    (doc, docRef, docIndex) => {
+      return (
+        <>
+          {scrollTriggerRef && docIndex === 0 && <div ref={scrollTriggerRef} />}
+          <KeylineCard ref={docRef}>
+            <Document
+              doc={doc}
+              editable={isEditable}
+              isTimeSeries={isTimeSeries}
+              copyToClipboard={copyToClipboard}
+              removeDocument={removeDocument}
+              replaceDocument={replaceDocument}
+              updateDocument={updateDocument}
+              openInsertDocumentDialog={openInsertDocumentDialog}
+            />
+          </KeylineCard>
+        </>
+      );
+    },
+    [
+      isEditable,
+      isTimeSeries,
+      scrollTriggerRef,
+      copyToClipboard,
+      openInsertDocumentDialog,
+      removeDocument,
+      replaceDocument,
+      updateDocument,
+    ]
+  );
+
+  return (
+    <VirtualList
+      items={docs}
+      renderItem={renderItem}
+      estimateItemInitialHeight={estimateDocumentInitialHeight}
+      rowGap={spacing[200]}
+      className={className}
```

Comment:
Looks like for the `list` and `json` view we're now showing the padding in a way where it doesn't reach the whole page, and the scrollbar inside of the padding.
Can we update how we apply those styles, `listAndJsonStyles`, so that the padding is instead inside the scrollable container? Then the documents reach the whole way to the bottom of the screen as well.

| Compass | branch |
| - | - |
| <img width="294" alt="Screenshot 2024-08-05 at 1 27 03 PM" src="https://github.com/user-attachments/assets/e8e97aa6-e493-40c3-81c4-ed51617fc374"> | <img width="386" alt="Screenshot 2024-08-05 at 1 26 41 PM" src="https://github.com/user-attachments/assets/96639e0c-2e45-4992-85ae-66ed88dbe77c"> |

Reply:
Just to confirm - is it that the documents don't reach all the way to the bottom or all the way to the right of Compass app (excluding padding-right)? Because I can see that the documents reach to the bottom up until the desired padding but yea the scrollbar getting in the way of padding-right is a real issue that troubles us also on the sidebar when the scroll is visible there. That's actually a problem with the library itself, with the way they render the scrollbar, its always inside the container and will render the scrollbar gutter as soon as it sees an overflow.

Reply:
> Just to confirm - is it that the documents don't reach all the way to the bottom or all the way to the right of Compass app (excluding padding-right)? 

I'd expect the documents to render up to the shell (when still scrolling). In main there isn't the space between the current document scrolled to and the shell (screenshot from this branch):
<img width="963" alt="Screenshot 2024-08-05 at 5 00 24 PM" src="https://github.com/user-attachments/assets/3659dacb-8d3c-47ee-8b4a-27d649b28ae6">

For the right, I expect the scrollbar to be on the farthest right, no space with the side of the app, with some spacing between the scrollbar and the document.

> That's actually a problem with the library itself, with the way they render the scrollbar, its always inside the container and will render the scrollbar gutter as soon as it sees an overflow.

Would removing the container padding and instead use margin on the documents work?

Reply:
Thank you for the suggestion Rhys 🙌  I could not get margins to work for us (likely because everything in the list container is calculated and margins / paddings invalidates those calculations) but there were [suggestions from the library](https://github.com/bvaughn/react-window#can-i-add-padding-to-the-top-and-bottom-of-a-list) which I followed here in this [1ecac93](https://github.com/mongodb-js/compass/pull/6069/commits/1ecac93e8bd0af2a8fd870f551904e0a07970c64). This fixes the problem you mentioned but as a side effect the scroll goes a little below the actual last item itself.

<img width="1183" alt="image" src="https://github.com/user-attachments/assets/0fa18e83-d035-4363-9baf-b6dc80fd2744">


So far my feeling is that it is unavoidable because of how we mimick adding padding to the scroll container which basically increasing the height of the container by the amount of padding.

Reply:
And this is how it looks for the right view of the scroll. As you can notice I purposely reduced the spacing between the scrollbar and the KeylineCard but that is to align the KeylineCard with the end of the top toolbar header contents. Let me know what you think?

<img width="1240" alt="image" src="https://github.com/user-attachments/assets/00c16604-5ac0-4e02-891f-3852a69abf32">

Reply:
Loots good !